### PR TITLE
Add support for rate limiting open ai calls

### PR DESCRIPTION
**Pull request by angrybayblade**
## Proposed changes

This PR

- Adds an open ai call tracker
- Adds a round at the beginning of the twitter scoring ABCI to check if the open ai calls has exceeded or not

## Fixes

If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce? (A **breaking change** is a fix or feature that would cause existing functionality and APIs to not work as expected.)
_Put an `x` in the box that applies_

- [ ] Non-breaking fix (non-breaking change which fixes an issue)
- [ ] Breaking fix (breaking change which fixes an issue)
- [ ] Non-breaking feature (non-breaking change which adds functionality)
- [ ] Breaking feature (breaking change which adds functionality)
- [ ] Refactor (non-breaking change which changes implementation)
- [ ] Messy (mixture of the above - requires explanation!)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `main` branch (left side). Also you should start your branch off our `main`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have locally run services that could be impacted and they do not present failures derived from my changes

### CODE REVIEW

File: packages/valory/skills/twitter_scoring_abci/behaviours.py [hunk 1]
```diff
@@ -79,6 +81,34 @@ def params(self) -> Params:
...
+    matching_round: Type[AbstractRound] = OpenAICallCheckRound
+
+    def async_act(self) -> Generator:
+        """Do the act, supporting asynchronous execution."""
+        with self.context.benchmark_tool.measure(self.behaviour_id).local():
+            current_time = cast(
+                SharedState, self.context.state
+            ).round_sequence.last_round_transition_timestamp.timestamp()
+            # Reset the window if the window expired before checking
+            self.params.openai_calls.reset(current_time=current_time)
```
Comment (0xArdi) [inline packages/valory/skills/twitter_scoring_abci/behaviours.py:96]: does this need to be synchronized? this seems like a local agent thing. I don't think its wrong, as long as the window length is larger than the round timeout.

**Review by 0xArdi — APPROVED** [review]

File: packages/valory/skills/twitter_scoring_abci/behaviours.py (hunk 1 above)
Comment (angrybayblade) [reply packages/valory/skills/twitter_scoring_abci/behaviours.py:96]: Yes, the window length is going to be way larger then the round timeouts

File: packages/valory/skills/twitter_scoring_abci/models.py [hunk 2]
```diff
@@ -38,9 +39,45 @@ class SharedState(BaseSharedState):
...
+    def __init__(
+        self,
+        openai_call_window_size: float,
+        openai_calls_allowed_in_window: int,
+    ) -> None:
+        """Initialize object."""
+        self._calls_made_in_window = 0
+        self._calls_allowed_in_window = openai_calls_allowed_in_window
+        self._call_window_size = openai_call_window_size
+        self._call_window_start = datetime.now().timestamp()
```
Comment (Adamantios) [inline packages/valory/skills/twitter_scoring_abci/models.py:54]: Not used.

```suggestion
```

File: packages/valory/skills/twitter_scoring_abci/models.py [hunk 3]
```diff
@@ -38,9 +39,45 @@ class SharedState(BaseSharedState):
...
+            return
+        self._tweets_made_in_window = 0
+        self._tweet_window_start = current_time
+
+
 class Params(BaseParams):
     """Parameters."""
 
+    _current_tweet_window_start: Optional[float]
+
```
Comment (Adamantios) [inline packages/valory/skills/twitter_scoring_abci/models.py:80]: Not used.

```suggestion
```

File: packages/valory/skills/twitter_scoring_abci/models.py [hunk 4]
```diff
@@ -60,6 +97,15 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
...
             "tweet_evaluation_round_timeout", kwargs, float
         )
+        self.openai_calls = OpenAICalls(
+            openai_call_window_size=self._ensure(
+                "openai_call_window_size", kwargs, float
+            ),
+            openai_calls_allowed_in_window=self._ensure(
+                "openai_calls_allowed_in_window", kwargs, int
+            ),
+        )
```
Comment (Adamantios) [inline packages/valory/skills/twitter_scoring_abci/models.py:107]: Not a common pattern. Parameters are meant to be frozen, you could create the `OpenAICalls` object in the shared state using two params: `openai_call_window_size` and `openai_calls_allowed_in_window`

File: packages/valory/skills/twitter_scoring_abci/behaviours.py [hunk 5]
```diff
@@ -79,6 +81,34 @@ def params(self) -> Params:
...
+            if self.params.openai_calls.max_tweets_reached():
+                content = OpenAICallCheckRound.API_CALL_EXCEEDED
+            else:
+                content = OpenAICallCheckRound.API_CALL_REMAINING
```
Comment (Adamantios) [inline packages/valory/skills/twitter_scoring_abci/behaviours.py:100]: Why not set the content to `None`? That way you would not need to override the round's `end_block` at all.

File: packages/valory/skills/twitter_scoring_abci/rounds.py [hunk 6]
```diff
@@ -89,6 +89,28 @@ def latest_mention_tweet_id(self) -> dict:
...
+        """Process the end of the block."""
+        if self.threshold_reached:
+            if self.most_voted_payload == self.API_CALL_EXCEEDED:
+                return self.synchronized_data, Event.API_ERROR
+            return self.synchronized_data, Event.DONE
+        if not self.is_majority_possible(
+            self.collection, self.synchronized_data.nb_participants
+        ):
+            return self.synchronized_data, Event.NO_MAJORITY
+        return None
```
Comment (Adamantios) [inline packages/valory/skills/twitter_scoring_abci/rounds.py:111]: All this is unnecessary if you use the `none_event` class attribute. See https://github.com/valory-xyz/IEKit/pull/78#discussion_r1305225837.

File: packages/valory/skills/twitter_scoring_abci/rounds.py [hunk 7]
```diff
@@ -89,6 +89,26 @@ def latest_mention_tweet_id(self) -> dict:
...
+
+    def end_block(self) -> Optional[Tuple[BaseSynchronizedData, Event]]:
+        """Process the end of the block."""
+        if self.threshold_reached:
+            return self.synchronized_data, Event.DONE
+        if not self.is_majority_possible(
+            self.collection, self.synchronized_data.nb_participants
+        ):
+            return self.synchronized_data, Event.NO_MAJORITY
+        return None
```
Comment (Adamantios) [inline packages/valory/skills/twitter_scoring_abci/rounds.py:109]: Still not needed :sweat_smile: 

Take a look at the base class. You also have done and no majority event attributes available.

**Review by Adamantios — APPROVED** [review]
Looks good, just a comment:

`OpenAICallCheckRound` could be a simple voting round in the future and corresponding behaviour would also be simplified. We just need to know if we are rate-limited or not.

**Pull request merged**